Properly handle ti state difference between executor and scheduler

ephraimbuddy opened the pull request:
When a task fails to start, the executor fails it and its state in
scheduler is queued while its state in executor is failed. Currently
we fail this task without retries to avoid getting stuck.

This PR changes this to only fail the task if the callback cannot be
executed. This ensures the task does not get stuck

closes: https://github.com/apache/airflow/issues/16625



---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

kaxil reviewed (changes requested):
We don't want to run any user code in Scheduler. That is why callbacks are currently run in the DAG file processor. Long term they should be run in the Worker

ephraimbuddy commented on tests/jobs/test_scheduler_job.py:
```diff
@@ -207,30 +207,32 @@ def test_process_executor_events(self, mock_stats_incr, mock_task_callback, dag_
         self.scheduler_job._process_executor_events(session=session)
         ti1.refresh_from_db()
-        assert ti1.state == State.SUCCESS
+        assert ti1.state == State.success_states
```
```suggestion
        assert ti1.state == State.SUCCESS
```

ephraimbuddy commented on tests/jobs/test_scheduler_job.py:
```diff
@@ -207,17 +207,18 @@ def test_process_executor_events(self, mock_stats_incr, mock_task_callback, dag_
             '<TaskInstance: test_process_executor_events.dummy_task 2016-01-01 00:00:00+00:00 [queued]> '
             'finished (failed) although the task says its queued. (Info: None) '
             'Was the task killed externally?',
+            task=task1,
```
I'm not sure why this is failing...will appreciate help here

ashb commented on airflow/utils/callback_requests.py:
```diff
@@ -59,8 +59,10 @@ def __init__(
         simple_task_instance: SimpleTaskInstance,
         is_failure_callback: Optional[bool] = True,
         msg: Optional[str] = None,
+        task: Optional[Any] = None,
```
We can't pass the task here -- this is a "message" that goes between processes, and by doing this it will require serialization of the task, but the task here in the scheduler is Serialized and so won't have access to any of the code etc.

Instead (if we need this sort of behaviour) we'll have to load the dag in the processor (which we already do) and _then_ execute it.

ephraimbuddy commented on airflow/jobs/scheduler_job.py:
```diff
@@ -596,13 +597,26 @@ def _process_executor_events(self, session: Session = None) -> int:
+                    session.merge(ti)
+                    session.commit()
+                    continue
+
```
Borrowed from 1.10.x

ashb commented on airflow/jobs/scheduler_job.py:
```diff
@@ -49,6 +49,7 @@
 from airflow.ti_deps.dependencies_states import EXECUTION_STATES
 from airflow.utils import timezone
 from airflow.utils.callback_requests import DagCallbackRequest, TaskCallbackRequest
+from airflow.utils.cli import get_dag
```
This function loads the actual DAG from the file on disk, which we can't do in the scheduler

Can't because Airflow shouldn't ever load DAGs in to any long running process which includes the scheduler. This is bad because it could mean a badly written dag file could bring down the scheduler (and right now the scheduler is isolated from this.

ashb commented on airflow/jobs/scheduler_job.py:
```diff
@@ -596,13 +597,24 @@ def _process_executor_events(self, session: Session = None) -> int:
                 self.log.error(msg, ti, state, ti.state, info)
 
+                try:
+                    get_dag(self.subdir, ti.dag_id)
```
It's not clear to me why we even need this? What's going on here?

ephraimbuddy replied:
When there's an error in parsing the dag, maybe due to resource constraints, the task callback is not run even though it's sent to the processor. The task gets stuck in queued. 
But when there's db operational error while trying to execute a task, it results in this error but the task retries.
One other thing we can do is add a message to inform the user that if there's a resource constraint that the task can get stuck. Then remove parsing of dags

ephraimbuddy committed:
Properly handle ti state difference between executor and scheduler

When a task fails to start, the executor fails it and the report
says that its state in scheduler is queued while its state in executor is failed.

Currently we fail this task without retries to avoid getting stuck.

This change modifies the above to only fail the task if there's no retries left

ephraimbuddy committed:
fixup! Properly handle ti state difference between executor and scheduler

kaxil commented on tests/jobs/test_scheduler_job.py:
```diff
@@ -207,7 +207,7 @@ def test_process_executor_events(self, mock_stats_incr, mock_task_callback, dag_
         ti1.refresh_from_db()
-        assert ti1.state == State.FAILED
+        assert ti1.state == State.QUEUED
         mock_task_callback.assert_called_once_with(
```
Hmmm..  I wonder if we can just handle failure in _process_executor_events itself if there are no callbacks. If there are no callbacks we probably don't need to send it to dag parsing process 

And we need a comment here, to say the state will remain in Queue here and will be set to failed in dag parsing process.

ephraimbuddy replied:
Ok. But I think we may miss the logs?

ashb replied:
If there are no callbacks what logs did you expect?

Oh, and sending it to the dag file parser process still won't produce any logs that in the "right" location to get picked up by the webserver.

ephraimbuddy replied:
The issue I face now is that to be able to check if there are still retries, we need the task object. `ti.is_eligible_to_retry` requires that `ti.task` exists

ephraimbuddy committed:
fixup! fixup! Properly handle ti state difference between executor and scheduler

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -1719,6 +1719,8 @@ def handle_failure(
             email_for_state = task.email_on_failure
         else:
+            if self.state == State.QUEUED:
+                self.try_number += 1
```
Need a comment explaining why we need this please

ashb commented on tests/models/test_taskinstance.py:
```diff
@@ -1683,6 +1683,21 @@ def test_handle_failure(self, create_dummy_dag, session=None):
+        assert ti.try_number == 1
+        ti.handle_failure("test queued ti", test_mode=True)
+        assert ti.state == State.UP_FOR_RETRY
+        assert ti.try_number == 3
```
Hmmm, this is right but my does it confusing!

ephraimbuddy replied:
Quiet confusing. I have put breakpoints to see how it's happening but no result. If I remove the increment of the try_numer, the try_number remains 1 after running the handle_failure

kaxil replied:
Yeah - `try_number` get's bumped to `2` in https://github.com/apache/airflow/blob/550ca654cd1b6c2e97a4174c3950a17093747880/airflow/models/taskinstance.py#L1722-L1723

and when `ti.try_number` is accessed here: https://github.com/apache/airflow/blob/550ca654cd1b6c2e97a4174c3950a17093747880/airflow/models/taskinstance.py#L467-L480

since it is not running and in  `UP_FOR_RETRY` state, the `try_number` is set to `3`.

ashb replied:
This might be a case where accessing `ti._try_number` would be less confusing.

ephraimbuddy committed:
Add comment

ephraimbuddy committed:
apply suggestions from code review

ephraimbuddy committed:
fixup! apply suggestions from code review

kaxil commented on airflow/jobs/scheduler_job.py:
```diff
@@ -521,14 +521,26 @@ def _process_executor_events(self, session: Session = None) -> int:
+                    dag = self.dagbag.get_dag(ti.dag_id)
+                    from airflow.models.baseoperator import BaseOperator
+
+                    task: BaseOperator = dag.get_task(ti.task_id)
```
```suggestion
                    task = dag.get_task(ti.task_id)
```

kaxil commented on airflow/models/taskinstance.py:
```diff
@@ -1693,6 +1693,7 @@ def handle_failure(
         Stats.incr(f'operator_failures_{task.task_type}', 1, 1)
         Stats.incr('ti_failures')
         if not test_mode:
+            self.dag_run = self.get_dagrun(session=session)
```
@ashb This had to be added otherwise it gave a `sqlalchemy.orm.exc.DetachedInstanceError` (`lazy load operation of attribute 'dag_run' cannot proceed`), specifically:

```
E           sqlalchemy.orm.exc.DetachedInstanceError: Parent instance <TaskInstance at 0x7fdfb98dd5d0> is not bound to a Session; lazy load operation of attribute 'dag_run' cannot proceed (Background on this error at: http://sqlalche.me/e/13/bhk3)

/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/strategies.py:720: DetachedInstanceError
```

any ideas?

ashb replied:
Not sure why this only cam up now, but this change looks okay

kaxil committed:
Apply suggestions from code review

ashb commented on airflow/jobs/scheduler_job.py:
```diff
@@ -521,14 +521,24 @@ def _process_executor_events(self, session: Session = None) -> int:
+                    dag = self.dagbag.get_dag(ti.dag_id)
+                    task = dag.get_task(ti.task_id)
+                except Exception as ex:
+                    self.log.exception("Marking task instance %s as failed. Reason: %s", ti, ex)
```
`log.exception` already includes the exception, so we don't need to include the message again.
```suggestion
                    self.log.exception("Marking task instance %s as %s", ti, state)
```

ashb commented on airflow/jobs/scheduler_job.py:
```diff
@@ -521,14 +521,24 @@ def _process_executor_events(self, session: Session = None) -> int:
+                    ti.set_state(state)
+                    continue
+                ti.task = task
+                if task.on_retry_callback or task.on_failure_callback:
```
@kaxil Is this right for serialized tasks?

kaxil replied:
Yeah, we currently store it as string:

```python
(Pdb) pp SerializedDAG.serialize_dag(dag)
{'_dag_id': 'test_process_executor_events_with_callback',
 '_task_group': {'_group_id': None,
                 'children': {'dummy_task': (<DagAttributeTypes.OP: 'operator'>,
                                             'dummy_task')},
                 'downstream_group_ids': [],
                 'downstream_task_ids': [],
                 'prefix_group_id': True,
                 'tooltip': '',
                 'ui_color': 'CornflowerBlue',
                 'ui_fgcolor': '#000',
                 'upstream_group_ids': [],
                 'upstream_task_ids': []},
 'dag_dependencies': [],
 'edge_info': {},
 'fileloc': '/test_path1/',
 'params': {},
 'schedule_interval': {<Encoding.TYPE: '__type'>: <DagAttributeTypes.TIMEDELTA: 'timedelta'>,
                       <Encoding.VAR: '__var'>: 86400.0},
 'start_date': 1451606400.0,
 'tasks': [{'_downstream_task_ids': [],
            '_inlets': [],
            '_is_dummy': True,
            '_outlets': [],
            '_task_module': 'airflow.serialization.serialized_objects',
            '_task_type': 'SerializedBaseOperator',
            'label': 'dummy_task',
            'on_failure_callback': '            task1 = '
                                   'DummyOperator(task_id=task_id_1, '
                                   'on_failure_callback=lambda x: '
                                   'print("hi"))\n',
            'pool': 'default_pool',
            'task_id': 'dummy_task',
            'template_ext': [],
            'template_fields': [],
            'template_fields_renderers': {},
            'ui_color': '#e8f7e4',
            'ui_fgcolor': '#000'}],
 'timezone': 'UTC'}
```

or 

```python
(Pdb) from airflow.serialization.serialized_objects import SerializedBaseOperator, SerializedDAG
(Pdb) pp SerializedDAG.serialize_dag(dag)
{'_dag_id': 'test_process_executor_events_with_callback',
 '_task_group': {'_group_id': None,
                 'children': {'dummy_task': (<DagAttributeTypes.OP: 'operator'>,
                                             'dummy_task')},
                 'downstream_group_ids': [],
                 'downstream_task_ids': [],
                 'prefix_group_id': True,
                 'tooltip': '',
                 'ui_color': 'CornflowerBlue',
                 'ui_fgcolor': '#000',
                 'upstream_group_ids': [],
                 'upstream_task_ids': []},
 'dag_dependencies': [],
 'edge_info': {},
 'fileloc': '/test_path1/',
 'params': {},
 'schedule_interval': {<Encoding.TYPE: '__type'>: <DagAttributeTypes.TIMEDELTA: 'timedelta'>,
                       <Encoding.VAR: '__var'>: 86400.0},
 'start_date': 1451606400.0,
 'tasks': [{'_downstream_task_ids': [],
            '_inlets': [],
            '_is_dummy': True,
            '_outlets': [],
            '_task_module': 'airflow.serialization.serialized_objects',
            '_task_type': 'SerializedBaseOperator',
            'label': 'dummy_task',
            'on_retry_callback': '            task1 = '
                                 'DummyOperator(task_id=task_id_1, '
                                 'on_retry_callback=lambda x: print("hi"))\n',
            'pool': 'default_pool',
            'task_id': 'dummy_task',
            'template_ext': [],
            'template_fields': [],
            'template_fields_renderers': {},
            'ui_color': '#e8f7e4',
            'ui_fgcolor': '#000'}],
 'timezone': 'UTC'}
(Pdb)
```

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -1719,6 +1720,11 @@ def handle_failure(
+                # We increase the try_number so as
+                # to fail the task if it fails to start
+                # after sometime
+                self.try_number += 1
```
```suggestion
                self._try_number += 1
```

ashb replied:
(This might not have any affect but is more correct -- as that way we don't have to reason about "has the try_number property already added 1 to this value or not")

kaxil committed:
fixup! Apply suggestions from code review
